src/components/shared/NewDownloadButton.jsx

```diff
@@ -0,0 +1,37 @@
+import React, { useState, useRef } from 'react';
+import Button from '@material-ui/core/Button';
+import DescriptionIcon from '@material-ui/icons/Description';
+import CircularProgress from '@material-ui/core/CircularProgress';
+
+export default function NewDownloadButton({ displayText, getData, fileName }) {
+  const [blobURL, setBlobURL] = useState('');
+  const [loading, setLoading] = useState(false);
+
+  const hiddenButton = useRef();
+
+  async function download() {
+    setLoading(true);
+
+    const response = await getData();
+    const blob = new Blob([JSON.stringify(response)], { type: 'application/json' });
+    setBlobURL(window.URL.createObjectURL(blob));
+    hiddenButton.current.click();
+
+    setLoading(false);
+  }
+
+  return (
+    <>
+      <Button
+        onClick={!loading && download}
+        startIcon={!loading && <DescriptionIcon />}
+        variant="contained"
+        size="large"
+        color="secondary"
+      >
+        {loading ? <CircularProgress /> : displayText}
+      </Button>
+      <a ref={hiddenButton} href={blobURL} download={fileName()} />
```

Comment:
I tried making this comment on the last review, but something must have gone wrong in Github. Is there an advantage to doing the download this way, with the <a> tag in the html? My preference is towards how we did this in the DownloadButton.jsx downloadJSONSimple function.

Reply:
This is really the same way as what is being done in the DownloadButton, the difference being that it is done programmatically in DownloadButton. I like this way better because we get to use a ref, which feels more React-esque than inserting something as a child of the body. The difference is all style though, the outcome is the exact same.

Reply:
I would like to do it the way we did in the original DownloadButton. That lets you remove the html, the blobURL state, the ref, and encapsulates all the download logic in the function. Also that fileName() function in the download prop is being run on every render currently.